### DISCUSSION TITLE
esbuild is unable to tree-shake rxjs under node platform

### DISCUSSION THREAD

**Discussion by Conaclos** [ReactiveX/rxjs#6785] [Report issues other than bug]
Hi!

I discovered that **esbuild** is not able to tree-shake **rxjs** package when bundling under node platform.
[**esbuild** priorizes exports condition 'node' over main-fields](https://esbuild.github.io/api/#how-conditions-work).

# Demonstration

```sh
$ cat index.js
export { Subject } from 'rxjs'

$ esbuild --bundle index.js --platform=node | wc -l
9654

$ # Specifying main-fields does not change the output
$ esbuild --bundle index.js --platform=node --main-fields=es2015,module,main | wc -l
9654

$ # browser bundle is lighter
$ esbuild --bundle index.js --platform=browser | wc -l
1161

```

Note that we need to bundle for node since our project and most of our dependencies imports node-specific modules.

# How to solve the issue?

**rxjs** should not use the condition "node".
It should use the condition "require":

```json
  ...
  "exports": {
    ".": {
      "require": "./dist/cjs/index.js",
      "es2015": ...,
      "default": ...
    },
    ....
```

instead of:

```json
  ...
  "exports": {
    ".": {
      "node": "./dist/cjs/index.js",
      "es2015": ...,
      "default": ...
    },
    ....
```

**Comment by Conaclos** [comment]
If the node condition is important, another possibility is to rely on a sub-condition to give more control to the end-user. For instance, it is possible to add the sub-condition `module` [*][1]:

```json
  ...
  "exports": {
    ".": {
      "node": {
        "module": "./dist/esm5/ajax/index.js",
        "default": "./dist/cjs/index.js"
      },
      "es2015": ...,
      "default": ...
    },
    ....
```

[1]: https://webpack.js.org/guides/package-exports/#providing-commonjs-and-esm-version-stateless

**Comment by benlesh** [comment]
Just for clarity: Why would tree-shaking be required/desirable on a server? Generally, "node platform" would indicate that you're building something that's not being shipped over the wire. What are the negative impacts, so I can understand them? Is it just initial load/parse time?

**Comment by Conaclos** [comment]
@benlesh
I maintains a package which enables to run a webrtc signaling server. I would like to distribute a single-file executable to run the server without requiring npm. Tree-shaking is useful in this case to reduce the file's size as much as possible.

**Reply by benlesh** [reply]
I'm not sure I'll have time in the near future to examine what we can do about this. But I can tell you that by version 8 we are going to have the main RxJS library, and then we're also going to publish packages that include individual things like observable or operators. In the meantime, you could import directly from under internal as a workaround. It's not recommended, but it's not going to break anytime during version 7.

**Reply by Conaclos** [reply]
Thanks for your time and your suggestion. I understand the time issue. I will explore your suggestion and other possible workarounds.
